token/core/zkatdlog/crypto/validator/validator_transfer.go

```diff
@@ -0,0 +1,187 @@
+	"github.com/hyperledger-labs/fabric-token-sdk/token/core/zkatdlog/crypto/transfer"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/interop/htlc"
+	"github.com/pkg/errors"
+)
+
+type Context struct {
+	PP                *crypto.PublicParams
+	Deserializer      driver.Deserializer
+	SignatureProvider driver.SignatureProvider
+	Signatures        [][]byte
+	InputTokens       []*token.Token
+	Action            *transfer.TransferAction
+	Ledger            driver.Ledger
+}
+
+type ValidateTransferFunc func(ctx *Context) error
+
+func TransferSignatureValidate(ctx *Context) error {
+	var tokens []*token.Token
+	var signatures [][]byte
+
+	inputs, err := ctx.Action.GetInputs()
+	if err != nil {
+		return errors.Wrapf(err, "failed to retrieve inputs to spend")
+	}
+	for i, in := range inputs {
+		logger.Debugf("load token [%d][%s]", i, in)
+		bytes, err := ctx.Ledger.GetState(in)
+		if err != nil {
+			return errors.Wrapf(err, "failed to retrieve input to spend [%s]", in)
+		}
+		if len(bytes) == 0 {
+			return errors.Errorf("input to spend [%s] does not exists", in)
+		}
+
+		tok := &token.Token{}
+		if err := tok.Deserialize(bytes); err != nil {
+			return errors.Wrapf(err, "failed to deserialize input to spend [%s]", in)
+		}
+		tokens = append(tokens, tok)
+		logger.Debugf("check sender [%d][%s]", i, view.Identity(tok.Owner).UniqueID())
+		verifier, err := ctx.Deserializer.GetOwnerVerifier(tok.Owner)
+		if err != nil {
+			return errors.Wrapf(err, "failed deserializing owner [%d][%s][%s]", i, in, view.Identity(tok.Owner).UniqueID())
+		}
+		logger.Debugf("signature verification [%d][%s][%s]", i, in, view.Identity(tok.Owner).UniqueID())
+		sigma, err := ctx.SignatureProvider.HasBeenSignedBy(tok.Owner, verifier)
+		if err != nil {
+			return errors.Wrapf(err, "failed signature verification [%d][%s][%s]", i, in, view.Identity(tok.Owner).UniqueID())
+		}
+		signatures = append(signatures, sigma)
+	}
+
+	ctx.InputTokens = tokens
+	ctx.Signatures = signatures
+
+	return nil
+}
+
+func TransferZKProofValidate(ctx *Context) error {
+	in := make([]*math.G1, len(ctx.InputTokens))
+	for i, tok := range ctx.InputTokens {
+		in[i] = tok.GetCommitment()
+	}
+
+	if err := transfer.NewVerifier(
+		in,
+		ctx.Action.GetOutputCommitments(),
+		ctx.PP).Verify(ctx.Action.GetProof()); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func TransferHTLCValidate(ctx *Context) error {
+	now := time.Now()
+
+	for i, in := range ctx.InputTokens {
+		owner, err := identity.UnmarshallRawOwner(in.Owner)
+		if err != nil {
+			return errors.Wrap(err, "failed to unmarshal owner of input token")
+		}
+		if owner.Type == htlc.ScriptType {
+			if len(ctx.InputTokens) != 1 || len(ctx.Action.GetOutputs()) != 1 {
+				return errors.Errorf("invalid transfer action: an htlc script only transfers the ownership of a token")
+			}
+
+			out := ctx.Action.GetOutputs()[0].(*token.Token)
+			if ctx.InputTokens[0].Data.Equals(out.Data) {
```

Comment:
This check should be removed as it is incorrect. ZKVerifier takes care of checking if the transfer of ownership is correct.

Reply:
where is this check done exactly?